README.md

```diff
@@ -185,7 +284,39 @@ and [development](https://docs.ansible.com/ansible/latest/dev_guide/index.html).
 * File headers and functions should have comments for their intent.
 
 ## Ansible Best Practices
-
+* Ansible variables use lazy evaluation. [more_info](https://github.com/ansible/ansible/issues/10374)
+* All tags should be namespaced/prefixed with the role name.
+* Use preferably the command module instead of the shell module. Even better, use a dedicated module, if it
+  exists. If not, see the [section](#check-mode-and-idempotency-issues) about idempotency and check mode and
+  make sure that you support them properly (your task will likely need options such as `changed_when:`
+  and maybe `check_mode:` ). Anytime `command` or `shell` modules are used, a comment in the code with
+  justificiation would help with future maintenance.
+* Beware of bare variables in `when`. Use '`| bool`' if the variable is supposed to hold a boolean
+  value. [more_info](https://github.com/ansible/ansible/issues/39414).
```

Comment:
Another technique is to use `| d(EMPTY)` where `EMPTY` is the "empty" value for the variable type e.g. if expecting a `dict`, use `| d({})` - note that this is redundant when the variable is always defined in `vars/*.yml` or `defaults/main.yml`, but it can be the shortest, simplest way to avoid the bare variables problem

Reply:
Thinking about it, the recommendation here is maybe not even entirely correct. `| bool` has the main function of proper conversion of user-supplied boolean values (not of working around bare variables) and can yield unexpected results (if one expects it to perform Python-style boolean evaluation). 
Does it work to just wrap the variable in `( )` ?

Reply:
yes, `| d()` will work, as long as you are only testing to see if a variable is defined and has a non empty value.  https://jinja.palletsprojects.com/en/2.11.x/templates/#default
```
default(value, default_value='', boolean=False)
```
to me, this means `undefined_var | d()` will evaluate to a value of type `string` and value `''`, the empty string.  If you are just testing like `when: undefined_var | d()` because you want to only execute the task if the variable is defined and non-empty, then this usage is ok.  But if you want to further filter the result, you'll need to use the empty value of the expected type.